packages/aws-cdk/lib/docker.ts

```diff
@@ -0,0 +1,208 @@
+import { ContainerImageAssetMetadataEntry } from '@aws-cdk/cx-api';
+import { CloudFormation } from 'aws-sdk';
+import crypto = require('crypto');
+import { ToolkitInfo } from './api/toolkit-info';
+import { debug, print } from './logging';
+import { shell } from './os';
+import { PleaseHold } from './util/please-hold';
+
+/**
+ * Build and upload a Docker image
+ *
+ * Permanently identifying images is a bit of a bust. Newer Docker version use
+ * a digest (sha256:xxxx) as an image identifier, which is pretty good to avoid
+ * spurious rebuilds. However, this digest is calculated over a manifest that
+ * includes metadata that is liable to change. For example, as soon as we
+ * push the Docker image to a repository, the digest changes. This makes the
+ * digest worthless to determe whether we already pushed an image, for example.
+ *
+ * As a workaround, we calculate our own digest over parts of the manifest that
+ * are unlikely to change, and tag based on that.
+ */
+export async function prepareContainerAsset(asset: ContainerImageAssetMetadataEntry, toolkitInfo: ToolkitInfo): Promise<CloudFormation.Parameter[]> {
+  debug(' 👑  Preparing Docker image asset:', asset.path);
+
+  const buildHold = new PleaseHold(` ⌛ Building Docker image for ${asset.path}; this may take a while.`);
+  try {
+    buildHold.start();
+
+    const command = ['docker',
+      'build',
+      '--quiet',
+      asset.path];
+    const imageId = (await shell(command, { quiet: true })).trim();
+    buildHold.stop();
+
+    const tag = await calculateImageDigest(imageId);
+
+    debug(` ⌛  Image has tag ${tag}, preparing ECR repository`);
+    const ecr = await toolkitInfo.prepareEcrRepository(asset.id, tag);
+
+    if (ecr.alreadyExists) {
+      debug(' 👑  Image already uploaded.');
+    } else {
+      // Login and push
+      debug(` ⌛  Image needs to be uploaded first.`);
+
+      await shell(['docker', 'login',
+        '--username', ecr.username,
+        '--password', ecr.password,
+        ecr.endpoint]);
+
+      const qualifiedImageName = `${ecr.repositoryUri}:${tag}`;
+      await shell(['docker', 'tag', imageId, qualifiedImageName]);
+
+      // There's no way to make this quiet, so we can't use a PleaseHold. Print a header message.
```

Comment:
Uh? Can't redirect all the IO's to nothin'?

Reply:
Also, I am fine with showing docker's progress. This can take an hour

Reply:
Yeah, it's not horrible to leave it on.